fdbrpc/RESTUtils.actor.cpp

```diff
@@ -0,0 +1,276 @@
+
+	knobMap["connection_pool_size"] = std::addressof(connection_pool_size);
+	knobMap["pz"] = std::addressof(connection_pool_size);
+	knobMap["secure_connection"] = std::addressof(secure_connection);
+	knobMap["sc"] = std::addressof(secure_connection);
+	knobMap["connect_tries"] = std::addressof(connect_tries);
+	knobMap["ct"] = std::addressof(connect_tries);
+	knobMap["connect_timeout"] = std::addressof(connect_timeout);
+	knobMap["cto"] = std::addressof(connect_timeout);
+	knobMap["max_connection_life"] = std::addressof(max_connection_life);
+	knobMap["mcl"] = std::addressof(max_connection_life);
+	knobMap["request_tries"] = std::addressof(request_tries);
+	knobMap["rt"] = std::addressof(request_tries);
+	knobMap["request_timeout_secs"] = std::addressof(request_timeout_secs);
+	knobMap["rtom"] = std::addressof(request_timeout_secs);
+}
+
+void RESTClientKnobs::set(const std::unordered_map<std::string, int>& knobSettings) {
+	TraceEvent trace = TraceEvent("RESTClient_SetKnobs");
+
+	for (const auto& itr : knobSettings) {
+		const auto& kItr = RESTClientKnobs::knobMap.find(itr.first);
+		if (kItr == RESTClientKnobs::knobMap.end()) {
+			trace.detail("RESTClient_InvalidKnobName", itr.first);
+			throw rest_invalid_rest_client_knob();
+		}
+		*(kItr->second) = itr.second;
+		trace.detail(itr.first.c_str(), itr.second);
+	}
+}
+
+std::unordered_map<std::string, int> RESTClientKnobs::get() const {
+	std::unordered_map<std::string, int> details = {
+		{ "connection_pool_size", connection_pool_size },
+		{ "secure_connection", secure_connection },
+		{ "connect_tries", connect_tries },
+		{ "connect_timeout", connect_timeout },
+		{ "max_connection_life", max_connection_life },
+		{ "request_tries", request_tries },
+		{ "request_timeout_secs", request_timeout_secs },
+	};
+
+	return details;
+}
+
+ACTOR Future<RESTConnectionPool::ReusableConnection> connect_impl(Reference<RESTConnectionPool> connectionPool,
+                                                                  RESTConnectionPoolKey connectKey,
+                                                                  bool isSecure,
+                                                                  int maxConnLife) {
+	auto poolItr = connectionPool->connectionPoolMap.find(connectKey);
+	if (poolItr == connectionPool->connectionPoolMap.end()) {
+		throw rest_connectpool_key_not_found();
+	}
+
+	while (!poolItr->second.empty()) {
+		RESTConnectionPool::ReusableConnection rconn = poolItr->second.front();
+		poolItr->second.pop();
+
+		if (rconn.expirationTime > now()) {
+			TraceEvent("RESTClient_ReusableConnection")
+			    .suppressFor(60)
+			    .detail("RemoteEndpoint", rconn.conn->getPeerAddress())
+			    .detail("ExpireIn", rconn.expirationTime - now());
+			return rconn;
+		}
+	}
+
+	state Reference<IConnection> conn =
+	    wait(INetworkConnections::net()->connect(connectKey.first, connectKey.second, isSecure));
+	wait(conn->connectHandshake());
```

Comment:
S3BlobStore connect also has a hook for initializing stuff post-connection instantiation for credentials. While we may not need a post-connection hook here explicitly, we again still probably want to make sure this works with the refactor there